File: ui/lib/api/demo/transport.dart
import 'package:fixnum/fixnum.dart';
import 'package:mizer/api/contracts/transport.dart';
import 'package:mizer/protos/transport.pb.dart';
import 'package:mizer/protos/transport.pbenum.dart';

class TransportDemoApi implements TransportApi {
  @override
  Future<void> setBPM(double bpm) {
    // TODO: implement setBPM
    throw UnimplementedError();
  }

  @override
  Future<void> setState(TransportState state) {
    // TODO: implement setState
    throw UnimplementedError();
  }

  @override
  Stream<Transport> watchTransport() {
    return Stream.periodic(Duration(milliseconds: 16), (frames) {
      int seconds = (frames / 60).floor();
      int minutes = (seconds / 60).floor();
      int hours = (minutes / 60).floor();

      return Transport(
        speed: 90,
        state: TransportState.Playing,
        timecode: Timecode(
          frames: Int64(frames - (seconds * 60)),
          seconds: Int64(seconds - (minutes * 60)),
          minutes: Int64(minutes - (hours * 60)),
          hours: Int64(hours),
        )
      );
    });
  }
}

File: crates/api/src/mappings/transport.rs
use mizer_clock::{ClockSnapshot, ClockState};

use crate::proto::transport::{Timecode, Transport, TransportState};

impl From<ClockSnapshot> for Transport {
    fn from(snapshot: ClockSnapshot) -> Self {
        Transport {
            speed: snapshot.speed,
            state: TransportState::from(snapshot.state) as i32,
            timecode: Some(snapshot.time.into()),
            fps: snapshot.fps,
        }
    }
}

impl From<mizer_clock::Timecode> for Timecode {
    fn from(timecode: mizer_clock::Timecode) -> Self {
        Self {
            frames: timecode.frames,
            hours: timecode.hours,
            minutes: timecode.minutes,
            seconds: timecode.seconds,
        }
    }
}

impl From<ClockState> for TransportState {
    fn from(state: ClockState) -> Self {
        match state {
            ClockState::Playing => Self::Playing,
            ClockState::Paused => Self::Paused,
            ClockState::Stopped => Self::Stopped,
        }
    }
}

impl From<TransportState> for ClockState {
    fn from(state: TransportState) -> Self {
        match state {
            TransportState::Playing => Self::Playing,
            TransportState::Paused => Self::Paused,
            TransportState::Stopped => Self::Stopped,
        }
    }
}
